packages/flutter/lib/src/widgets/heroes.dart

```diff
@@ -897,3 +899,47 @@ class HeroController extends NavigatorObserver {
     return toHero.child;
   };
 }
+
+/// Enables or disables [Hero]es in the widget subtree.
+///
+/// When [enabled] is false, all Hero widgets in this subtree will not be
+/// considered for Hero animations.
+///
+/// When [enabled] is true (the default), Hero widgets may be involved in a
+/// Hero animation, like normal.
+class HeroMode extends InheritedWidget {
```

Comment:
It should not be an inherited widget. The inherited widget will cause unnecessary rebuild when its flag changes.

I think It should be ProxyWidget

Reply:
Only a few widgets below inherit ProxyWidget.

- ParentDataWidget
- InheritedWidget

I am still not convinced if ProxyWidget is appropriate for HeroMode. I think that it seems too fundamental and unsuitable for inheritance.

If we use ProxyWidget, we also need to implement Element. It seems overkill to me for what we are trying to do.

If updateShouldNotify always returns false, then it can avoid unnecessary rebuild?

Reply:
no, inherited widget adds additional overhead to to the element tree(all the inherited widget will be cached on each element in the tree, see Element._inheritedWidgets), and it is not something we should use if we don't needs its inherited functionality.

Although proxy widget is not used as often as other other type of widget, it seems to be the best choice for this case without adding additional overhead

Reply:
I understood that InheritedWidget has overhead.

How about StatelessWidget? I think StatelessWidget is good for this.

Reply:
stateless widget will introduce an additional call to the build method, if you use proxy widget you can directly return the child. Although it may not be a big deal, but why not perfect it?

Reply:
If we use ProxyWidget, we need to implement HeroModeElement. It is big deal than introducing build method.

The following widgets inherit StatelessWidget and do `Widget build(BuildContext context) => child;`, not inherit ProxyWidget.

- NotificationListener
- KeyedSubtree
- PageStorage
- PreferredSize

Reply:
yes you will need to implement HeroModeElement which you can extend from the ComponentElement, in that class you just need to be sure to implement build and update method. You can take a look at StatelessElement

Reply:
Is using ProxyWidget faster than using StatelessWidget? Are there any other benefits?

OK, I will try!

Reply:
If we implement HeroModeElement in the same way as StatelessElement, we can use StatelessWidget instead of ProxyWidget. I do not understand why to use ProxyWidget. What is the difference between HeroModeElement and StatelessElement?

Reply:
Not much difference except the build method of HeroModeElement does not do a `widget.build(context)`, instead it just return widget.child. This save us a function call.

Reply:
StatelessElement uses Element._dirty, HeroModeElement can not use the variable because it class is in the outside. And it is not good to implement HeroModeElement in widgets/framework.dart.

Reply:
I see, then it seems like the only way to use the proxy widget will be overriding the StatelessElement and override the build method.

I agree this may make the code looks weird and probably not worth the effort to the save one build method. yeah, let's use stateless widget, what you have looks good.